src/contracts/BCoWPool.sol

```diff
@@ -0,0 +1,201 @@
+    SOLUTION_SETTLER_DOMAIN_SEPARATOR = ISettlement(_cowSolutionSettler).domainSeparator();
+    VAULT_RELAYER = ISettlement(_cowSolutionSettler).vaultRelayer();
+  }
+
+  /**
+   * @notice Once this function is called, it will be possible to trade with
+   * this AMM on CoW Protocol.
+   * @param appData Trading is enabled with the appData specified here.
+   */
+  function enableTrading(bytes32 appData) external onlyController {
+    bytes32 _appDataHash = keccak256(abi.encode(appData));
+    appDataHash = _appDataHash;
+    emit TradingEnabled(_appDataHash, appData);
+  }
+
+  /**
+   * @notice Disable any form of trading on CoW Protocol by this AMM.
+   */
+  function disableTrading() external onlyController {
+    appDataHash = NO_TRADING;
+    emit TradingDisabled();
+  }
+
+  /**
+   * @notice Restricts a specific AMM to being able to trade only the order
+   * with the specified hash.
+   * @dev The commitment is used to enforce that exactly one AMM order is
+   * valid when a CoW Protocol batch is settled.
+   * @param orderHash the order hash that will be enforced by the order
+   * verification function.
+   */
+  function commit(bytes32 orderHash) external {
+    if (msg.sender != address(SOLUTION_SETTLER)) {
+      revert CommitOutsideOfSettlement();
+    }
+    assembly ("memory-safe") {
+      tstore(COMMITMENT_SLOT, orderHash)
+    }
+  }
+
+  /**
+   * @inheritdoc IERC1271
+   * @dev this function reverts if the order hash does not match the current commitment
+   */
+  function isValidSignature(bytes32 _hash, bytes memory signature) external view returns (bytes4) {
+    (GPv2Order.Data memory order) = abi.decode(signature, (GPv2Order.Data));
+
+    if (appDataHash != keccak256(abi.encode(order.appData))) revert AppDataDoNotMatchHash();
+    bytes32 orderHash = order.hash(SOLUTION_SETTLER_DOMAIN_SEPARATOR);
+    if (orderHash != _hash) revert OrderDoesNotMatchMessageHash();
+
+    if (orderHash != commitment()) {
+      revert OrderDoesNotMatchCommitmentHash();
+    }
+
+    verify(order);
+
+    // A signature is valid according to EIP-1271 if this function returns
+    // its selector as the so-called "magic value".
+    return this.isValidSignature.selector;
+  }
+
+  function commitment() public view returns (bytes32 value) {
+    assembly ("memory-safe") {
+      value := tload(COMMITMENT_SLOT)
+    }
+  }
+
+  /**
+   * @notice This function checks that the input order is admissible for the
+   * constant-product curve for the given trading parameters.
+   * @param order `GPv2Order.Data` of a discrete order to be verified.
+   */
+  function verify(GPv2Order.Data memory order) public view {
+    Record memory inRecord = _records[address(order.sellToken)];
+    Record memory outRecord = _records[address(order.buyToken)];
+
+    if (!inRecord.bound || !inRecord.bound) {
+      revert BPool_TokenNotBound();
+    }
+    if (order.validTo < block.timestamp + MAX_ORDER_DURATION) {
+      revert BCoWPool_OrderExpired();
```

Comment:
this is wrong, please look at CoW Constant Product pool
https://github.com/cowprotocol/cow-amm/blob/039c6524b2ddc032446597dfa073d021d1a5152c/src/ConstantProduct.sol#L345C1-L350C1

the requirement should be oposite, and "expired" is not the intended error, we should test that the order is only valid up to a certain point in time, the expiration of the order should be handled by CoW